src/backend/columnar/columnar_metadata.c

```diff
@@ -1331,13 +1292,14 @@ ReadMetapage(RelFileNode relfilenode, bool missingOk)
 
 		if (nblocks != 0)
 		{
-			metapageBuffer = ReadFromSmgr(relation, 0, sizeof(ColumnarMetapage));
+			metapage = palloc(sizeof(ColumnarMetapage));
+			ColumnarMetapageRead(relation, metapage);
```

Comment:
I'm trying to keep allocations out of columnar_storage.c because it's low-level and some callers will want to do their own allocation. For instance, `columnar_metadata.c:ReadMetadata()` was reading and then copying out of the buffer.